internal/alb/sg/association.go

```diff
@@ -0,0 +1,259 @@
+		LbArn:    association.LbArn,
+	}
+	err = controller.lbAttachmentController.Reconcile(lbSGAttachment)
+	if err != nil {
+		return lbSG, fmt.Errorf("failed to reconcile managed LoadBalancer securityGroup attachment due to %s", err.Error())
+	}
+	return lbSG, nil
+}
+
+func (controller *associationController) reconcileManagedInstanceSG(association *Association, lbSG *SecurityGroup) error {
+	instanceSGName := controller.namer.NameInstanceSG(association.LbID)
+	instanceSG := &SecurityGroup{
+		GroupName: &instanceSGName,
+		InboundPermissions: []*ec2.IpPermission{
+			{
+				IpProtocol: aws.String("tcp"),
+				FromPort:   aws.Int64(0),
+				ToPort:     aws.Int64(65535),
+				UserIdGroupPairs: []*ec2.UserIdGroupPair{
+					{
+						GroupId: lbSG.GroupID,
+					},
+				},
+			},
+		},
+	}
+	err := controller.sgController.Reconcile(instanceSG)
+	if err != nil {
+		return fmt.Errorf("failed to reconcile managed Instance securityGroup due to %s", err.Error())
+	}
+	instanceSGAttachment := &InstanceAttachment{
+		GroupID: *instanceSG.GroupID,
+		Targets: association.Targets,
+	}
+	err = controller.instanceAttachmentController.Reconcile(instanceSGAttachment)
+	if err != nil {
+		return fmt.Errorf("failed to reconcile managed Instance securityGroup attachment due to %s", err.Error())
+	}
+	return nil
+}
+
+func (controller *associationController) deletedManagedSGs(association *Association) error {
+	err := controller.deleteManagedInstanceSG(association)
+	if err != nil {
+		return fmt.Errorf("failed to delete managed Instance securityGroup due to %s", err.Error())
+	}
+
+	err = controller.deleteManagedLbSG(association)
+	if err != nil {
+		return fmt.Errorf("failed to delete managed LoadBalancer securityGroup due to %s", err.Error())
+	}
+	return nil
+}
+
+func (controller *associationController) deleteManagedLbSG(association *Association) error {
+	lbSGName := controller.namer.NameLbSG(association.LbID)
+	lbSGID, err := controller.findSGIDByName(lbSGName)
+	if err != nil {
+		return err
+	}
+	if lbSGID == nil {
+		return nil
+	}
+	lbSGAttachment := &LbAttachment{
+		GroupIDs: []string{*lbSGID},
+		LbArn:    association.LbArn,
+	}
+	err = controller.lbAttachmentController.Delete(lbSGAttachment)
+	if err != nil {
+		return err
+	}
+	lbSG := &SecurityGroup{
+		GroupID: lbSGID,
+	}
+	err = controller.sgController.Delete(lbSG)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (controller *associationController) deleteManagedInstanceSG(association *Association) error {
+	instanceSGName := controller.namer.NameInstanceSG(association.LbID)
+	instanceSGID, err := controller.findSGIDByName(instanceSGName)
+	if err != nil {
+		return err
+	}
+	if instanceSGID == nil {
+		return nil
+	}
+	instanceSGAttachment := &InstanceAttachment{
+		GroupID: *instanceSGID,
+	}
+	err = controller.instanceAttachmentController.Delete(instanceSGAttachment)
+	if err != nil {
+		return err
+	}
+	instanceSG := &SecurityGroup{
+		GroupID: instanceSGID,
```

Comment:
Why `instanceSGID` is used as value here by used as pointer in
```
instanceSGAttachment := &InstanceAttachment{
		GroupID: *instanceSGID,
	}
```
Should we make it consistent?

Reply:
instanceSGID i retrieved maybe nil, so it retrieved as *string. (Hope we got optional support in golang to better express the idea) 
however, in the API of my code, i deliberately make the inputs to be string if it's required and cannot be nil. 

I may have named the variable in `old styles` like pInstanceSGID, but it's unnecessary with modern IDE support 😄